### PR TITLE
Day19 - DinhVo - Service, Broadcast Receiver, Notification

### PR DESCRIPTION
**Pull request by at-dinhvo**
#### Related issues (Optional)
Task #100

#### Description (Required)
- [x] Play a list of songs ( online ).
- [x] Have function: Next - Previous - Play - Shuffle - Auto Next.
- [x] When press Lockscreen -> Show notification

### CODE REVIEW

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        if (mMediaPlayer != null) {
+            mMediaPlayer.release();
+        }
+        stopSelf();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/day20/adapter/MusicAdapter.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    public void onBindViewHolder(ItemViewHolder holder, int position) {
+        holder.mTvSong.setText(mSongs.get(position).getName());
+        holder.mTvSinger.setText(mSongs.get(position).getSinger());
+        loadImage(mSongs.get(position).getImage(), holder.mImgMusic);
+    }
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/adapter/MusicAdapter.java]: ERROR: [PMD-CPD] 19 lines duplicated.
app/src/main/java/vn/asiantech/internship/drawer/ui/feed/FeedPagerAdapter.java

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+            Log.e(TAG, "onDestroy: " + 1);
+        }
+        if (mMediaPlayer != null) {
+            mMediaPlayer.release();
+            Log.e(TAG, "onDestroy: ");
+        }
+        stopSelf();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+            Log.e(TAG, "onDestroy: " + 1);
+        }
+        if (mMediaPlayer != null) {
+            mMediaPlayer.release();
+            Log.e(TAG, "onDestroy: ");
+        }
+        mLength = 0;
+        mSong = null;
+        unregisterReceiver(mBroadcastReceiver);
+        stopSelf();
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: @at-dinhvo 
Please fix this warning

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+                } else if (intent.getAction().equals(String.valueOf(TelephonyManager.CALL_STATE_IDLE))) {
+                    mIsPause = false;
+                    resumeMusic();
+                }
+            }
+        }
+    };
+
+    public MusicService() {
+        // constructor
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: `constructor` to `No-op`

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+    private void sendDuration() {
+        Intent durationIntent = new Intent();
+        durationIntent.setAction(DURATION);
+        durationIntent.putExtra(KEY_TIME_INT, mMediaPlayer.getDuration() / 1000);
+        sendBroadcast(durationIntent);
+    }
+
+    private void showNotification(String song, String singer, int duration, int currentTime) {
+        RemoteViews views = new RemoteViews(getPackageName(), R.layout.custom_notification);
+//        RemoteViews bigViews = new RemoteViews(getPackageName(), R.layout.custom_notification_expanse);
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: Please remove all `comment - code` if you don't you it

File: app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+            Toast.makeText(getContext(), "Position is wrong!", Toast.LENGTH_SHORT).show();
+        }
+    }
+
+    private void addEvents() {
+        mImgBtnPlay.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View view) {
+                Intent intentPlay = new Intent();
+                if (mIsPause) { // if music is pausing
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java]: Please change `if music is pausing` to `If music is pausing`
And please also check other one

File: app/src/main/res/anim/rotation.xml [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+    <rotate
+        android:duration="5000"
+        android:fromDegrees="0"
+        android:interpolator="@android:anim/linear_interpolator"
+        android:pivotX="50%"
+        android:pivotY="50%"
+        android:repeatCount="infinite"
+        android:repeatMode="restart"
+        android:toDegrees="3600" />
+</set>
```
Comment (at-vinhhuynh) [inline app/src/main/res/anim/rotation.xml]: add a break line

File: app/src/main/res/drawable/bg_button_next.xml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:drawable="@drawable/ic_skip_next_cyan_a700_48dp" android:state_pressed="true" />
+    <item android:drawable="@drawable/ic_skip_next_white_48dp" />
+</selector>
```
Comment (at-vinhhuynh) [inline app/src/main/res/drawable/bg_button_next.xml]: add a break line

File: app/src/main/res/drawable/bg_button_pause.xml [hunk 10]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:drawable="@drawable/ic_pause_cyan_a700_48dp" android:state_pressed="true" />
+    <item android:drawable="@drawable/ic_pause_white_48dp" />
+</selector>
```
Comment (at-vinhhuynh) [inline app/src/main/res/drawable/bg_button_pause.xml]: add a break line

File: app/src/main/res/drawable/bg_button_play.xml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:drawable="@drawable/ic_play_arrow_cyan_a700_48dp" android:state_pressed="true" />
+    <item android:drawable="@drawable/ic_play_arrow_white_48dp" />
+</selector>
```
Comment (at-vinhhuynh) [inline app/src/main/res/drawable/bg_button_play.xml]: add a break line

File: app/src/main/res/drawable/bg_button_previous.xml [hunk 12]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="utf-8"?>
+<selector xmlns:android="http://schemas.android.com/apk/res/android">
+    <item android:drawable="@drawable/ic_skip_previous_cyan_a700_48dp" android:state_pressed="true" />
+    <item android:drawable="@drawable/ic_skip_previous_white_48dp" />
+</selector>
```
Comment (at-vinhhuynh) [inline app/src/main/res/drawable/bg_button_previous.xml]: add a break line

File: app/src/main/res/drawable/bg_notification.xml [hunk 13]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<bitmap xmlns:android="http://schemas.android.com/apk/res/android"
+    android:src="@drawable/notify_panel_notification_icon_bg"
+    android:tileMode="repeat" />
```
Comment (at-vinhhuynh) [inline app/src/main/res/drawable/bg_notification.xml]: add a break line

File: app/src/main/res/layout/custom_notification_expanse.xml [hunk 14]
```diff
@@ -0,0 +1,111 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    android:id="@+id/notificationbg"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/custom_notification_expanse.xml]: `notificationbg` to `rlNotification`

File: app/src/main/res/layout/custom_notification_expanse.xml [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    <ImageView
+        android:id="@+id/imgNotifiMusic"
+        android:layout_width="@dimen/notification_expanded_height"
+        android:layout_height="@dimen/notification_expanded_height"
+        android:layout_alignParentBottom="true"
+        android:contentDescription="@null"
+        android:scaleType="centerCrop" />
+
+    <LinearLayout
+        android:id="@+id/buttons"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/custom_notification_expanse.xml]: This is `LinearLayout ` please change it's `id` follow convention

File: app/src/main/res/layout/custom_notification_expanse.xml [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+            android:layout_weight="1.0"
+            android:background="?android:selectableItemBackground"
+            android:contentDescription="@null"
+            android:padding="10.0dip"
+            android:scaleType="fitCenter"
+            android:src="@drawable/apollo_holo_dark_play" />
+
+        <ImageButton
+            android:id="@+id/imgBtnNotifiNext"
+            android:layout_width="0.0dip"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/custom_notification_expanse.xml]: Please check and replace `dip`, `px` to `dp`

File: app/src/main/res/layout/custom_notification_expanse.xml [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        android:layout_height="wrap_content"
+        android:layout_alignParentRight="true"
+        android:layout_alignParentTop="true"
+        android:background="?android:selectableItemBackground"
+        android:contentDescription="@null"
+        android:padding="8.0dip"
+        android:src="@drawable/apollo_holo_dark_notifiation_bar_collapse" />
+
+    <LinearLayout
+        android:id="@+id/textArea"
```
Comment (at-vinhhuynh) [inline app/src/main/res/layout/custom_notification_expanse.xml]: `textArea` to `lnTextArea`

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 18]
```diff
@@ -0,0 +1,354 @@
...
+    @Override
+    public void onDestroy() {
+        super.onDestroy();
+        if (mCountDownTimer != null) {
+            mCountDownTimer.cancel();
+        }
+        if (mMediaPlayer != null) {
+            mMediaPlayer.release();
+        }
+        unregisterReceiver(mBroadcastReceiver);
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: WARNING: [CallSuperLast] super should be called at the end of the method
https://pmd.github.io/pmd-5.5.1/pmd-java/rules/java/android.html#CallSuperLast

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 19]
```diff
@@ -0,0 +1,342 @@
...
+        startForeground(111, notification);
+    }
+
+    @Override
+    public void onCreate() {
+        super.onCreate();
+        filterReceiver();
+    }
+
+    private void filterReceiver(){
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 20]
```diff
@@ -0,0 +1,349 @@
...
+    private int shuffleMusic() {
+        Random random = new Random();
+        int ran;
+        do {
+            ran = random.nextInt(mSongs.size() - 1);
+        } while (ran == mCurrentPosition);
+        return ran;
+    }
+
+    private void exitMusic(){
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 21]
```diff
@@ -0,0 +1,349 @@
...
+            notification = new Notification.Builder(this).build();
+            notification.contentView = views;
+            notification.flags = Notification.FLAG_ONGOING_EVENT;
+            notification.icon = R.drawable.ic_music_note_white_48dp;
+            notification.contentIntent = pendingIntent;
+        }
+        startForeground(111, notification);
+    }
+
+    private void filterReceiver(){
```
Comment (at-dev-support-01) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/vn/asiantech/internship/day20/model/Action.java [hunk 22]
```diff
@@ -0,0 +1,22 @@
+package vn.asiantech.internship.day20.model;
+
+/**
+ * Created by at-dinhvo on 07/07/2017.
+ */
+public enum Action {
+    PLAY("play"), PAUSE("pause"),
+    RESUME("resume"), NEXT("next"),
+    PREVIOUS("previous"), STOP("stop"),
+    SHUFFLE("shuffle"), AUTONEXT("auto_next"),
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/model/Action.java]: `AUTONEXT` to `AUTO_NEXT`

File: app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java [hunk 23]
```diff
@@ -0,0 +1,349 @@
...
+import java.util.List;
+import java.util.Random;
+
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.day20.model.Action;
+import vn.asiantech.internship.day20.model.Song;
+import vn.asiantech.internship.day20.ui.MusicActivity;
+import vn.asiantech.internship.day20.ui.MusicFragment;
+
+import static vn.asiantech.internship.day20.ui.MusicFragment.SONG_NEXT;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/service/MusicService.java]: remove `import-static`

File: app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+import android.widget.Toast;
+
+import java.util.ArrayList;
+
+import de.hdodenhof.circleimageview.CircleImageView;
+import vn.asiantech.internship.R;
+import vn.asiantech.internship.day20.model.Action;
+import vn.asiantech.internship.day20.model.Song;
+
+import static vn.asiantech.internship.day20.service.MusicService.DURATION;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java]: remove `import-static`

File: app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+    private SeekBar mSeekBar;
+    private TextView mTvMusicTime;
+    private TextView mCurrentTime;
+
+    private Animation mAnimation;
+    private boolean mIsPause;
+    private boolean mIsPlaying;
+    private boolean mIsShuffle;
+    private boolean mIsAutoNext;
+    private ArrayList<Song> mSongs;
```
Comment (at-vinhhuynh) [inline app/src/main/java/vn/asiantech/internship/day20/ui/MusicFragment.java]: `ArrayList` to `List`

**Comment by at-tienhoang** [comment]
lgtm

**Pull request merged**